### PR TITLE
[enterprise-metrics] Add Grafana Enterprise Metrics chart

### PR DESCRIPTION
**Pull request by jdbaldry**
This PR adds the Helm chart for [Grafana Metrics Enterprise](https://grafana.com/products/metrics-enterprise/). The files are migrated from the existing [Grafana Metrics Enterprise Helm chart](https://github.com/grafana/metrics-enterprise-helm-chart) which was itself orginally a fork of the [Cortex Helm chart](https://github.com/cortexproject/cortex-helm-chart/).

### CODE REVIEW

**Comment by jdbaldry** [comment]
Looks like some linting work is needed :) marking as draft while I fix that up.

**Comment by jdbaldry** [comment]
Putting this on hold until the next Metrics Enterprise release which will remove the requirement for a separate bootstrapping process and will simplify fixing the `ct install` tests.

File: .helmdocsignore [hunk 1]
```diff
@@ -2,3 +2,5 @@ charts/fluent-bit
 charts/grafana
 charts/loki
 charts/loki-stack
+charts/metrics-enterprise
+charts/promtail
```
Comment (unguiculus) [inline .helmdocsignore:6]: The promtail chart has been updated and should not be ignored

Comment (jdbaldry) [reply .helmdocsignore:6]: Ah yeah apologies. I think that was a rebase artifact. I'll make sure to clean that up

**Comment by jdbaldry** [comment]
@unguiculus I'm getting closer to this being ready for review. The chart installation first depends on the user creating a Kubernetes Secret containing a Grafana Metrics Enterprise license. The license wouldn't necessarily need to be valid for the service to start up in a degraded (no enterprise features) way but the Secret would need to exist. Is there a way of having set up steps using the chart testing tool or should I look at having the Secret optionally created as part of the Helm chart?

File: .helmdocsignore (hunk 1 above)
Comment (unguiculus) [reply .helmdocsignore:6]: No problem, I just came across this and thought I'd drop a quick comment.

**Comment by unguiculus** [comment]
> Is there a way of having set up steps using the chart testing tool or should I look at having the Secret optionally created as part of the Helm chart?
You can't have any setup steps. In any case, the chart should have an option to create the secret.

**Comment by jdbaldry** [comment]
cc. @suitupalex who is the main user of this chart right now. List of known breaking changes from the v0.2.4 release:
- [FEATURE]: bootstrap job no longer required
All components perform a bootstrap routine at startup. `tokengen` can be used to for admin token generation.
- [FEATURE]: NGINX replaced by gateway
Service name has changed from `<name>-metrics-enterprise` to `<name>-metrics-enterprise-gateway`
- [FEATURE]: license secret automatically created by Chart 
By default an invalid license is used, to use your own license, use something like `--set-file license.contents=./license.jwt`

**Review by suitupalex — COMMENTED** [review]
I think this is good to go. The only thing we might want to consider is the compactor resource defaults - thoughts @jdbaldry ?

**Review by suitupalex — DISMISSED** [review]

**Review by suitupalex — DISMISSED** [review]
LGTM - just gave it another fresh install and everything's working!

**Comment by jdbaldry** [comment]
@unguiculus I'm having a little difficulty getting the CI tests to pass. There is a little inter component dependency which is usually resolved within a couple container restarts but I think that may be preventing the CI from passing. Would it be possible to get this merged in so that the chart can be downloaded using the Grafana helm repository and then I track the clean up work in Issues?

**Review by unguiculus — CHANGES_REQUESTED** [review]
Charts have to pass CI in order to get merged. We should make no exception here. Please do the cleanup first. You will have to do that anyways going forward.

**Comment by jdbaldry** [comment]
Sure, I understand that. Am I correct in thinking that for the Helm chart to consider the install successful all the resources have to have been created and no Pods should have had any container restarts in the process?

**Comment by unguiculus** [comment]
A Helm install needs to be successful.

**Comment by jdbaldry** [comment]
@unguiculus, sorry to @ you again. I have noticed that the `minio-make-bucket-job` Pod is never scheduled in the CI event logs and is the reason for the install timing out as other components depend on the minio buckets being present. When I install this chart into my dev GKE cluster or when @suitupalex installs it to the EKS cluster she is using it works fine but not in CI at the moment.

At the command line I was able to install the chart successfully with:
```console
$ kind create cluster
$ helm install --generate-name ./charts/metrics-enterprise --values ./charts/metrics-enterprise-ci/test-yaml
$ kubectl get events --sort-by firstTimestamp | grep minio
```

Would you have any pointers for how to diagnose this further?

**Comment by unguiculus** [comment]
@unguiculus I'll have a look.

**Comment by jdbaldry** [comment]
Thanks! I think I may just be coming up against a wall with the need for the minio-make-bucket-job to run as soon as possible and with Helm having a Kind based ordering. Unfortunately, right now there is a hard object store dependency so I can't skip using Minio. Also unfortunately, the Minio chart doesn't expose initContainers or overriding the command so I'm probably I'm going to need to replace the Minio deployment for CI with my own one that makes the directories and then execs the Minio process.

I did try using hook-weight to make the job run sooner but had no luck, it still works consistently locally but I just don't see the Job created in CI.

**Comment by unguiculus** [comment]
Looking into this now. I can reproduce the issue locally but don't know why yet. If I run `ct install`, the hook doesn't seem to be triggered which is strange because under the hood `ct` just calls `helm install`. I'll have to dig deeper.

**Comment by jdbaldry** [comment]
I really appreciate you taking a look. Please let me know if there is anything I can do to help you out. I too looked at the `ct` code but wasn't sure where to go when I saw that it was just performing a helm install.

**Comment by unguiculus** [comment]
The problem is the `--wait` flag that's passed by `ct`. Helm waits until everything is installed before running post-install hook. I would have expected dependencies to be considered independently which doesn't seem to be the case.

See https://github.com/helm/helm/blob/master/pkg/action/install.go#L348-L364

**Comment by unguiculus** [comment]
This behavior is actually documented. Helm waits before post-install hooks are executed.
See https://helm.sh/docs/topics/charts_hooks/#hooks-and-the-release-lifecycle

As a workaround you could create an init container that creates the bucket. You could re-use the same script that's used by MinIO's post-install hook. The configmap containing the script is created.

**Comment by jdbaldry** [comment]
Thanks @unguiculus, you're totally right:

> The library loads the resulting resources into Kubernetes. Note that if the --wait flag is set, the library will wait until all resources are in a ready state and will not run the post-install hook until they are ready.

This is the bit that is biting me.

I think I may have to fork the Minio chart or just add Minio templates directly to this helm chart because they don't expose initContainer configuration and the chart is in a code freeze.

**Comment by unguiculus** [comment]
> I think I may have to fork the Minio chart or just add Minio templates directly to this helm chart because they don't expose initContainer configuration and the chart is in a code freeze.

Why? Have you seen what I had suggested?

**Comment by jdbaldry** [comment]
Ah do you mean an initContainer in one or all of the metrics-enterprise resources? That could work fine. I was getting hung up on the initContainer needing to be part of the Minio resources.

**Comment by unguiculus** [comment]
Yes, just an init container in one of your resources. As the configmap from the MinIO chart that contains the script for the bucket creation is successfully created, you can just mount that and run it.

**Comment by jdbaldry** [comment]
:tada: :tada: :tada: :tada: :tada: :tada: :tada: :tada:

**Comment by jdbaldry** [comment]
@unguiculus Really appreciate your help with this awkward PR.

I think we are good to go now for the initial addition. Is there anything else you would like me to add?

File: charts/enterprise-metrics/templates/_helpers.tpl [hunk 2]
```diff
@@ -0,0 +1,61 @@
+{{/* vim: set filetype=mustache: */}}
+{{/*
+Expand the name of the chart.
+*/}}
+{{- define "metrics_enterprise.name" -}}
```
Comment (unguiculus) [inline charts/enterprise-metrics/templates/_helpers.tpl:5]: Helper templates should use the name of the chart.
```suggestion
{{- define "enterprise-metrics.name" -}}
```

**Review by unguiculus — CHANGES_REQUESTED** [review]
I quickly went over it and have a few remarks. Please apply the inline comments throughout the chart.

Given this is a new chart, I wonder why it doesn't follow the official [label recommendations](https://kubernetes.io/docs/concepts/overview/working-with-objects/common-labels/). It seems you didn't use `helm create`. `helm create` was updated to follow these recommendations a long time ago, and it has since been best practice to use these labels. You'd also get helper templates for labels and selector labels which should be used. I'd really suggest you update the chart accordingly. Have a look at `loki-distributed` or `promtail` charts for reference.

Please use camel case consistently for values instead of a mixture of snake case and camel case. Camel case is commonly used for Helm charts. Please follow that practice.

I'd suggest you prefer `with` over `if` where possible. In some places you do, in others  you don't. Please make that consistent.

I'd encourage you to use `helm-docs`. This is already done for `loki-distributed`, `loki-canary`, and `promtail` charts. You document your `values.yaml` and get the README for free. You could, of course, do that in a follow-up PR.

Don't configure resources by default. Reasonable suggestions should be commented out.

I see you added alternative values files but don't seem to reference them anywhere. IMO, it would be better to document options or things to consider for environments of various sizes in the README. At least, you should link to them.

File: charts/enterprise-metrics/templates/_helpers.tpl (hunk 2 above)
Comment (jdbaldry) [reply charts/enterprise-metrics/templates/_helpers.tpl:5]: Good catch, I missed this during find and replace, thanks!

File: charts/enterprise-metrics/values.yaml [hunk 3]
```diff
@@ -0,0 +1,1084 @@
...
+    prometheus.io/scrape: 'true'
+    prometheus.io/port: 'http-metrics'
+
+  nodeSelector: {}
+  affinity: {}
+  annotations: {}
+
+  # DEPRECATED: use persistentVolume.subPath instead
+  persistence:
+    subPath:
```
[the remark this answers is not in the document]
Comment (jdbaldry) [reply charts/enterprise-metrics/values.yaml:224]: Most of these irregularities come from the [Cortex Helm chart](https://github.com/cortexproject/cortex-helm-chart) where this is forked from originally. The rest are me not yet knowing Helm best practices :D

In the future, I am considering a chart refactor to make this an umbrella chart that uses the Cortex Helm chart directly but I need to prove that the additional scope of the Cortex helm chart won't introduce more of a maintenance burden than it's worth.

[the remark this answers is not in the document]
Comment (jdbaldry) [reply charts/enterprise-metrics/values.yaml:224]: I've removed the `persistence` configuration

File: charts/enterprise-metrics/Chart.yaml [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+  - name: memcached
+    alias: memcached-queries
+    version: 3.2.3
+    repository: https://charts.helm.sh/stable
+    condition: memcached-queries.enabled
+  - name: memcached
+    alias: memcached-metadata
+    version: 3.2.3
+    repository: https://charts.helm.sh/stable
+    condition: memcached-metadata.enabled
```
[the remark this answers is not in the document]
Comment (jdbaldry) [reply charts/enterprise-metrics/Chart.yaml:25]: Yeah I have an issue on the old chart for this but haven't gotten around to making and testing those changes yet.

File: charts/enterprise-metrics/templates/admin-api-dep.yaml [hunk 5]
```diff
@@ -0,0 +1,160 @@
...
+metadata:
+  annotations:
+    {{- toYaml .Values.admin_api.annotations | nindent 4 }}
+  labels:
+    app: {{ template "metrics_enterprise.name" . }}-admin-api
+    chart: {{ template "metrics_enterprise.chart" . }}
+    release: {{ .Release.Name }}
+    heritage: {{ .Release.Service }}
+  name: {{ template "metrics_enterprise.fullname" . }}-admin-api
+  namespace: {{ .Release.Namespace }}
```
[the remark this answers is not in the document]
Comment (jdbaldry) [reply charts/enterprise-metrics/templates/admin-api-dep.yaml:12]: I've removed all the set namespaces

File: charts/enterprise-metrics/values.yaml [hunk 6]
```diff
@@ -0,0 +1,1084 @@
...
+  extraArgs:
+    {}
+
+  # Pod Labels
+  podLabels: {}
+
+  # Pod Annotations
+  podAnnotations:
+    prometheus.io/scrape: 'true'
+    prometheus.io/port: 'http-metrics'
```
[the remark this answers is not in the document]
Comment (jdbaldry) [reply charts/enterprise-metrics/values.yaml:216]: I've removed these from the chart.

**Comment by jdbaldry** [comment]
> Given this is a new chart, I wonder why it doesn't follow the official label recommendations. It seems you didn't use helm create. 
It's not a new chart which is why it wasn't created with `helm create`. I'll add the labels in to conform to best practice.


> Please use camel case consistently for values instead of a mixture of snake case and camel case. Camel case is commonly used for Helm charts. Please follow that practice.

The snake_case for objects like `store_gateway` was inherited from the Cortex helm chart and I would like to avoid diverging from that chart where I can. There are also users of this chart in the current form that will be broken by this change. I am definitely for consistency and following best practice so I would like to make an Issue to appropriately clean this up in the future, ideally in step with cleaning up the Cortex helm chart too to maintain compatibility.

> I'd suggest you prefer with over if where possible. In some places you do, in others you don't. Please make that consistent.

I appreciate the `with` syntactic sugar but would prefer to defer those changes for the same reasons as for the snake_case discussion above.

> I'd encourage you to use helm-docs. This is already done for loki-distributed, loki-canary, and promtail charts. You document your values.yaml and get the README for free. You could, of course, do that in a follow-up PR.

Yeah, I'd really like to start using helm-docs. I'll do that in a follow-up PR.

> Don't configure resources by default. Reasonable suggestions should be commented out.

I believe this relates to the default annotations which I have now removed.

> I see you added alternative values files but don't seem to reference them anywhere. IMO, it would be better to document options or things to consider for environments of various sizes in the README. At least, you should link to them.

Yes, I need to add docs for these.

**Review by unguiculus — APPROVED** [review]
Sorry for the delay. I'm just approving this as is now hoping for follow-up PRs that will consider what I had suggested.

**Comment by jdbaldry** [comment]
@unguiculus No problem whatsoever. I really appreciate the work you are putting into this community project. I will make sure to track the proposed changes and get to fixing them up as soon as I can.

**Pull request merged**